### DISCUSSION TITLE
Applying motion correction shifts to additional channel

### DISCUSSION THREAD

**Discussion by pbl007** [flatironinstitute/CaImAn#1238] [Q&A]
Hello,
I would deeply appreciated if someone could  tell me how to obtain the "shifts" computed for each frame and apply those to a different channel (e.g. Channel 1 Ca2+ data, Channel 2 cell-specific marker).
Thanks in advance!

**Comment by EricThomson** [comment]
Please see demo_seeded_CNMF.ipynb which shows how to do this. It's a demo I need to work on some more, but it should contain what you need.

**Reply by pbl007** [reply]
Thanks Eric for pointing to that notebook. It is very informative!

I see from the documentation of `apply_shift_movie` that the shifts are read directly from mc.pw_rigid. My intention is to apply the shift after running CNMF to an additional channel. Can I create the mc object from the mmap file? That will set me going with this for sure. Thanks again!

Pablo

**Reply by pbl007** [reply]
Hi. Still stuck on this one... will be great to know how to obtain the shifts for a session that already run to completion and be able to apply those to another channel.
Thanks!

**Reply by EricThomson** [reply]
@pbl007 sorry I missed your additional questions -- thanks for following up.

Caiman is not great about discussing saving motion correction parameters. In the typical use case it saves the motion corrected movie for subsequent source separation, but in this use case that is obviously not useful! 

I will look into it today.

**Comment by EricThomson** [comment]
@pbl007 The canonical use case is to run motion correction on a structural channel (e.g., red), and then apply the shifts to a different channel (green). Unfortunately, one weakness in Caiman is we don't currently have a good way to save these motion correction parameters after a run to save them for later use on the different channel/movie. 

This is obviously a serious problem in the code, and we should fix it! You cannot reconstruct it from the memmapped file, as that just saves the motion corrected file. What we need to do is create a save function, like we have for the CNMF object. This should be done for every major estimator in the caiman pipeline. I don't know why it doesn't exist for the motion correction object, but I'll open an issue about this now.

Sorry I didn't appreciate the nature of your question before. This is something that has bothered me for a while and is something that should be fixed.

**Reply by pbl007** [reply]
Thanks Eric for looking into this issue. Hope issue could be solved soon.

**Comment by EricThomson** [comment]
Added issue for this: #1246 

In the meantime, currently the shifts are not saved, so you need to apply the discovered shifts to the second channel straight away. Sorry about this!

**Reply by pbl007** [reply]
No problem! I'll change our pipeline in the meantime.
